### ISSUE TITLE
ecto :naive_datetime type not preserved when used in :map field

### ISSUE THREAD

**Issue by InteNs** [elixir-ecto/ecto#2355]
### Environment

* Elixir version (elixir -v):
```
Erlang/OTP 20 [erts-9.0.4] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Elixir 1.5.1
```
* Database and version (PostgreSQL 9.4, MongoDB 3.2, etc.):
```
postgres:latest
```
* Ecto version (mix deps):
```
{:phoenix_ecto,   "~> 3.2"}
```
* Database adapter and version (mix deps):
```
{:postgrex,       ">= 0.0.0"}
```
* Operating system:
```
macOS High Sierra (10.13.2)
```

### Current behavior
naive_datetime type not saved for ecto :map type (postgres jsonb)

```ex
iex(0)> {:ok, action} = Repo.insert(%Action{execute_params: %{date: ~N[2010-10-10 10:00:00]}})
{:ok,
 %Actionator.Schemas.Action{__meta__: #Ecto.Schema.Metadata<:loaded, "actions">,
  compensate_params: nil, execute_params: %{date: ~N[2010-10-10 10:00:00]},
  id: "fe006996-43b8-4d29-bbb7-0f911f426c17", name: nil, order: nil,
  saga: #Ecto.Association.NotLoaded<association :saga is not loaded>,
  saga_id: nil, status: nil, steps: nil}}
iex(1)> action = Repo.get(Action, action.id)
%Actionator.Schemas.Action{__meta__: #Ecto.Schema.Metadata<:loaded, "actions">,
 compensate_params: nil, execute_params: %{"date" => "2010-10-10T10:00:00"},
 id: "fe006996-43b8-4d29-bbb7-0f911f426c17", name: nil, order: nil,
 saga: #Ecto.Association.NotLoaded<association :saga is not loaded>,
 saga_id: nil, status: nil, steps: nil}
```
### Expected behavior
I expect the standard ecto types to be preserved when using :map type

```ex
iex(0)> {:ok, action} = Repo.insert(%Action{execute_params: %{date: ~N[2010-10-10 10:00:00]}})
{:ok,
 %Actionator.Schemas.Action{__meta__: #Ecto.Schema.Metadata<:loaded, "actions">,
  compensate_params: nil, execute_params: %{date: ~N[2010-10-10 10:00:00]},
  id: "fe006996-43b8-4d29-bbb7-0f911f426c17", name: nil, order: nil,
  saga: #Ecto.Association.NotLoaded<association :saga is not loaded>,
  saga_id: nil, status: nil, steps: nil}}
iex(1)> action = Repo.get(Action, action.id)
%Actionator.Schemas.Action{__meta__: #Ecto.Schema.Metadata<:loaded, "actions">,
 compensate_params: nil, execute_params: %{"date" => ~N[2010-10-10 10:00:00]},
 id: "fe006996-43b8-4d29-bbb7-0f911f426c17", name: nil, order: nil,
 saga: #Ecto.Association.NotLoaded<association :saga is not loaded>,
 saga_id: nil, status: nil, steps: nil}
```
### Ecto Schema:
```ex
  @primary_key {:id, :binary_id, autogenerate: true}

  schema "actions" do
    field :name,              :string
    field :order,             :integer
    field :execute_params,    :map
    field :compensate_params, :map
    field :status,            :string
    field :steps,             {:array, :string}
    belongs_to :saga, Saga
  end

```

**Comment by ericmj** [comment]
If the map is stored in a json(b) column then this type information is lost. Since JSON doesn't have datetimes as a type we have to convert it to a string. If you use an embedded schema and declare the field to be of `naive_datetime` type then we can decode the string back to a struct.

**Comment by josevalim** [comment]
You could try to convert it back to a NaiveDateTime if you use a JSON decoder/encoder that would attempt to do so but, for the reasons @ericmj mentioned, it is a bit error prone. In other words, the map fields delegate to the JSON library, so it is up to whatever it does.

**Comment by InteNs** [comment]
@ericmj  I would love to use embedded schemas, the only problem is that we never know what params are stored in the`execute_params` field.

every "Action" has it's own static permitted parameters so we can define an embedded schema for each action.

So we got the `name` attribute which may contain: `Actions.Activity.RealizeActivity'
(this is how I implemented polymorphism for different actions)
this refers to a module where we could define the embedded schema

is there any way we could define an `embeds_one` based on this `name` attribute?
